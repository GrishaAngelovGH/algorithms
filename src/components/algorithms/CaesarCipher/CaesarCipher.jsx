import { useState } from 'react'

import Button from 'react-bootstrap/Button'
import ButtonGroup from 'react-bootstrap/ButtonGroup'

import AlgorithmDescription from 'components/AlgorithmDescription'
import WaveBackground from 'components/WaveBackground'

import TextArea from './TextArea'

const rot13 = str => {
	const alphabet = [...'abcdefghijklmnopqrstuvwxyz']
	const words = str.split(' ')

	return words.map(word => {
		return [...word].map(v => {
			const index = (alphabet.indexOf(v) + 13) % alphabet.length
			return alphabet[index]
		}).join('')
	}).join(' ')
}

const CaesarCipher = () => {
	const [word, setWord] = useState('')
	const [encryptedWord, setEncryptedWord] = useState('')

	const handleTextAreaChange = value => {
		setWord(value)
		setEncryptedWord(rot13(value))
	}

	const handleSwap = () => {
		setWord(encryptedWord)
		setEncryptedWord(rot13(encryptedWord))
	}

	const handleReset = () => {
		setWord('')
		setEncryptedWord('')
	}

	return (
		<WaveBackground>
			<div className='row g-0 justify-content-center text-center p-3'>
				<div className='col-md-12'>
					<h3 className='text-white'>Caesar Cipher Algorithm</h3>

					<div className='row justify-content-center align-items-center p-2'>
						<div className='col-lg-6'>
							<h4 className='text-white'>Raw Input</h4>
							<TextArea value={word} onChange={handleTextAreaChange} />
						</div>

						<div className='row justify-content-center align-items-center'>
							<div className='col-lg-6 p-3'>
								<ButtonGroup size='lg' className='mb-2'>
									<Button onClick={handleSwap}>Swap</Button>
									<Button onClick={handleReset}>Reset</Button>
								</ButtonGroup>
							</div>
						</div>

						<div className='row justify-content-center align-items-center'>
							<div className='col-lg-6'>
								<h4 className='text-white'>Encryped Output</h4>
								<TextArea value={encryptedWord} disabled />
							</div>
						</div>
					</div>

					<div className='row justify-content-center m-3'>
						<div className='col-md-7'>
							<AlgorithmDescription>
								Caesar Cipher algorithm will replace each letter in given word
								with a different one, a fixed number of places down the alphabet ( 13 in this case )
							</AlgorithmDescription>
						</div>
					</div>
				</div>
			</div>
		</WaveBackground>
	)
}

export default CaesarCipher