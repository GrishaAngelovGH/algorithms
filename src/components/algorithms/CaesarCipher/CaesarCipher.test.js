import { render } from '@testing-library/react'

import CaesarCipher from './CaesarCipher'

test('should render component', () => {
    const container = render(<CaesarCipher />)

    expect(container).toMatchSnapshot()
})