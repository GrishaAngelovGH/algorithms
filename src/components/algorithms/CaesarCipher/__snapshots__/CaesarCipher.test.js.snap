// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="waveWrapper waveAnimation"
      >
        <div
          class="waveWrapperInner bgTop"
        >
          <div
            class="wave waveTop"
          />
        </div>
        <div
          class="waveWrapperInner bgMiddle"
        >
          <div
            class="wave waveMiddle"
          />
        </div>
        <div
          class="waveWrapperInner bgBottom"
        >
          <div
            class="wave waveBottom"
          />
        </div>
      </div>
      <div
        class="wave-content"
      >
        <div
          class="row g-0 justify-content-center text-center p-3"
        >
          <div
            class="col-md-12"
          >
            <h3
              class="text-white"
            >
              Caesar Cipher Algorithm
            </h3>
            <div
              class="row justify-content-center align-items-center p-2"
            >
              <div
                class="col-lg-6"
              >
                <h4
                  class="text-white"
                >
                  Raw Input
                </h4>
                <form
                  action=""
                  class="paper"
                  method="get"
                >
                  <textarea
                    class="text"
                    name="text"
                    placeholder="Enter text"
                  />
                </form>
              </div>
              <div
                class="row justify-content-center align-items-center"
              >
                <div
                  class="col-lg-6 p-3"
                >
                  <div
                    class="mb-2 btn-group btn-group-lg"
                    role="group"
                  >
                    <button
                      class="btn btn-primary"
                      type="button"
                    >
                      Swap
                    </button>
                    <button
                      class="btn btn-primary"
                      type="button"
                    >
                      Reset
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="row justify-content-center align-items-center"
              >
                <div
                  class="col-lg-6"
                >
                  <h4
                    class="text-white"
                  >
                    Encryped Output
                  </h4>
                  <form
                    action=""
                    class="paper"
                    method="get"
                  >
                    <textarea
                      class="text"
                      disabled=""
                      name="text"
                      placeholder=""
                    />
                  </form>
                </div>
              </div>
            </div>
            <div
              class="row justify-content-center m-3"
            >
              <div
                class="col-md-7"
              >
                <h4
                  class="text-white bg-primary p-2 rounded"
                >
                  Caesar Cipher algorithm will replace each letter in given word with a different one, a fixed number of places down the alphabet ( 13 in this case )
                </h4>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="waveWrapper waveAnimation"
    >
      <div
        class="waveWrapperInner bgTop"
      >
        <div
          class="wave waveTop"
        />
      </div>
      <div
        class="waveWrapperInner bgMiddle"
      >
        <div
          class="wave waveMiddle"
        />
      </div>
      <div
        class="waveWrapperInner bgBottom"
      >
        <div
          class="wave waveBottom"
        />
      </div>
    </div>
    <div
      class="wave-content"
    >
      <div
        class="row g-0 justify-content-center text-center p-3"
      >
        <div
          class="col-md-12"
        >
          <h3
            class="text-white"
          >
            Caesar Cipher Algorithm
          </h3>
          <div
            class="row justify-content-center align-items-center p-2"
          >
            <div
              class="col-lg-6"
            >
              <h4
                class="text-white"
              >
                Raw Input
              </h4>
              <form
                action=""
                class="paper"
                method="get"
              >
                <textarea
                  class="text"
                  name="text"
                  placeholder="Enter text"
                />
              </form>
            </div>
            <div
              class="row justify-content-center align-items-center"
            >
              <div
                class="col-lg-6 p-3"
              >
                <div
                  class="mb-2 btn-group btn-group-lg"
                  role="group"
                >
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Swap
                  </button>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
            </div>
            <div
              class="row justify-content-center align-items-center"
            >
              <div
                class="col-lg-6"
              >
                <h4
                  class="text-white"
                >
                  Encryped Output
                </h4>
                <form
                  action=""
                  class="paper"
                  method="get"
                >
                  <textarea
                    class="text"
                    disabled=""
                    name="text"
                    placeholder=""
                  />
                </form>
              </div>
            </div>
          </div>
          <div
            class="row justify-content-center m-3"
          >
            <div
              class="col-md-7"
            >
              <h4
                class="text-white bg-primary p-2 rounded"
              >
                Caesar Cipher algorithm will replace each letter in given word with a different one, a fixed number of places down the alphabet ( 13 in this case )
              </h4>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
